azure-iot-hub/tests/test_iothub_amqp_client.py

```diff
@@ -34,58 +23,109 @@
     "expiryTimeUtc": 1584727659,
     "messageId": 42,
 }
-fake_empty_prop = {}
-"""----Shared fixtures----"""
-
+fake_hostname = "hostname.mytest.net"
+fake_shared_access_key_name = "test_key_name"
+fake_shared_access_key = "Zm9vYmFy"
+fake_token_scope = "https://token.net/.default"
+fake_token = "fasdjkhg"
+fake_token_expiry = 1584727659
 
-@pytest.fixture(scope="function", autouse=True)
-def mock_uamqp_SendClient(mocker):
-    mock_uamqp_SendClient = mocker.patch.object(uamqp, "SendClient")
-    return mock_uamqp_SendClient
 
+class SharedAmqpClientTests(object):
+    @pytest.fixture(autouse=True)
+    def mock_uamqp_SendClient(self, mocker):
+        mock_uamqp_SendClient = mocker.patch.object(uamqp, "SendClient")
+        return mock_uamqp_SendClient
 
-@pytest.mark.describe("IoTHubAmqpClient - Amqp Client Connections")
-class TestIoTHubAmqpClient(object):
-    @pytest.mark.it("Send Message To Device")
-    def test_send_message_to_device(self, mocker, mock_uamqp_SendClient):
-        iothub_amqp_client = IoTHubAmqpClient(
-            fake_hostname, fake_shared_access_key_name, fake_shared_access_key
-        )
-        iothub_amqp_client.send_message_to_device(fake_device_id, fake_message, fake_app_prop)
+    @pytest.mark.it("Sends messages with application properties using the uamqp SendClient")
+    def test_send_message_to_device_app_prop(self, client, mock_uamqp_SendClient):
+        client.send_message_to_device(fake_device_id, fake_message, fake_app_prop)
         amqp_client_obj = mock_uamqp_SendClient.return_value
 
         assert amqp_client_obj.queue_message.call_count == 1
```

Comment:
Also, checking that the mocked methods were called itself doesn't mean much unless you show that the specific message was also passed in, and that the properties were set on that message. This test doesn't actually really test anything other than that you can pass in a message and some properties - it doesn't tell you what happens to them.

For this test you'd probably want something like the following asserts:

```python
# Message was queued
assert amqp_client_obj.queue_message.call_count == 1
msg_obj = amqp_client_obj.queue_message.call_args[0][0]
# Message was configured with properties and destination
assert msg_obj.message_content = fake_message    # .message_content might be the wrong name for this data, I don't know uamqp offhand
assert msg_obj.properties.to == "/devices/{}/messages/devicebound".format(fake_device_id)   # again, not sure where this is actually stored
assert msg_obj.application_properties == fake_app_prop
# Message was sent
assert amqp_client_obj.send_all_messages.call_count == 1
assert amqp_client_obj.send_all_messages.call_args == mocker.call(close_on_done=False)
```

You miiiiiiight even want to take it a bit farther and show that the message send doesn't happen until AFTER the message is queued to keep things really airtight